### PR TITLE
add autofix for `PYI055`

### PR DESCRIPTION
**Pull request by diceroll123**
## Summary

Adds autofix for `PYI055`

Questions:
- is it intended that only the code within the `def func():` gets fixed? 🤔 I feel like some of the top-level variables would also be affected.
- should `w: builtins.type[int] | builtins.type[str] | builtins.type[complex]` turn into `w: builtins.type[int | str | complex]`?
  - or `w: type[int | str | complex]`?

## Test Plan

`cargo test`, and manually.

### CODE REVIEW

File: crates/ruff_linter/src/rules/flake8_pyi/rules/unnecessary_type_union.rs [hunk 1]
```diff
@@ -74,15 +92,83 @@ pub(crate) fn unnecessary_type_union<'a>(checker: &mut Checker, union: &'a Expr)
...
+                                    .collect(),
+                                ctx: ExprContext::Load,
+                                range: TextRange::default(),
+                            })),
+                            ctx: ExprContext::Load,
+                            range: TextRange::default(),
+                        })),
+                        ctx: ExprContext::Load,
+                        range: TextRange::default(),
+                    }))
```
Comment (diceroll123) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/unnecessary_type_union.rs:163]: is there a shorthand for any of this?

Comment (charliermarsh) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/unnecessary_type_union.rs:163]: No, it's bad 😞

Comment (diceroll123) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/unnecessary_type_union.rs:163]: That is _NOT_ what I wanted to hear! 🤣 Thanks anyways.

Comment (charliermarsh) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/unnecessary_type_union.rs:163]: Hopefully clear that I mean the abstractions are bad, and not your code :)

Comment (diceroll123) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/unnecessary_type_union.rs:163]: 🤣 of course.

...but why not both?

**Comment by charliermarsh** [comment]
I think `w: builtins.type[int] | builtins.type[str] | builtins.type[complex]` should turn into `w: builtins.type[int | str | complex]`.

Is your first question still relevant? It looks like stuff outside of the `def` is being flagged in the snapshot.

**Comment by diceroll123** [comment]
> I think `w: builtins.type[int] | builtins.type[str] | builtins.type[complex]` should turn into `w: builtins.type[int | str | complex]`.
> 
> Is your first question still relevant? It looks like stuff outside of the `def` is being flagged in the snapshot.

mmm, that answers my question! I'll dig into that. Thanks!

**Comment by charliermarsh** [comment]
On second thought, it's probably not worth the complexity. But we should ensure that `checker.is_builtin("type")` before adding the autofix.

**Comment by diceroll123** [comment]
> On second thought, it's probably not worth the complexity. But we should ensure that `checker.is_builtin("type")` before adding the autofix.

```rust
checker
        .semantic()
        .resolve_call_path(subscript.value.as_ref())
        .is_some_and(|call_path| matches!(call_path.as_slice(), ["" | "builtins", "type"]))
```

is this code starting on line 83 not doing that? or is it not explicit enough?

**Comment by charliermarsh** [comment]
It's not quite sufficient, because that would also match `builtins.type`. I just added a guard to check the builtin explicitly prior to performing the fix.

**Pull request merged**